bug: handle xml with namespaces for validations #1427

javiersvg opened the pull request:
This PR updates all places where xpath is used to be able to handle xml with name spaces.

marcingrzejszczak commented on spring-cloud-contract-verifier/src/test/java/org/springframework/cloud/contract/verifier/util/ContractVerifierUtilTest.java:
```diff
@@ -0,0 +1,127 @@
+		//When
+		String value = ContractVerifierUtil.valueFromXPath(parsedXml, "/customer/email/text()");
+		//Then
+		assertEquals("customer@test.com", value);
```
Can we use AssertJ here instead of junit assertions?

javiersvg replied:
Done

javiersvg replied:
done

marcingrzejszczak commented on spring-cloud-contract-verifier/src/test/java/org/springframework/cloud/contract/verifier/builder/XmlBodyVerificationBuilderTest.java:
```diff
@@ -0,0 +1,33 @@
+		builder.addXmlResponseBodyCheck(blockBuilder, xml, matchers, xml, true);
+		//Then
+		String test = blockBuilder.toString();
+		assertTrue(test.contains("DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();"));
```
Can we use AssertJ instead of JUnit assertions?

javiersvg replied:
Done

javiersvg replied:
done

marcingrzejszczak commented on spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/XmlBodyVerificationBuilder.java:
```diff
@@ -45,7 +45,9 @@ void addXmlResponseBodyCheck(BlockBuilder blockBuilder, Object responseBody,
 			String responseString) {
 		Arrays.asList(
-				"DocumentBuilder documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder()",
+				"DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance()",
```
I don't remember if you've already added those but have you added necessary imports for these classes?

javiersvg replied:
I checked and the imports are already there.

javiersvg commented:
I updated the PR with the requested changes. Also fixed checkstyle errors and added a class to parse the xml based on the DOMNamespaceContext from the JDK. In the first change set I implemented the management of the namespaces with an instance of the DOMNamespaceContext class present in the jdk which worked in the Groovy code botth for tests and running the maven plugin, but once I had to extent the implementation to Java code the classes were unable to reach that class an therefore I had to implement a NamespaceContext implementation of my own, for which I created the class present in the PR with the same name.
After uploading the latest PR I see thee build failing due to:
```
[ERROR] Errors: 
[ERROR]   StubRunnerJUnit5ExtensionCustomPortTests » FatalStartup java.lang.RuntimeExcep...
[INFO] 
[ERROR] Tests run: 198, Failures: 0, Errors: 1, Skipped: 0
```
This test belongs to the stub runner module which I have not modified and locally does not fail.
Would you help me find the reason behind this problem?

marcingrzejszczak commented:
I've reran the build, maybe some race condition happened.

marcingrzejszczak commented on spring-cloud-contract-verifier/src/test/java/org/springframework/cloud/contract/verifier/builder/XmlBodyVerificationBuilderTest.java:
```diff
@@ -0,0 +1,37 @@
+		builder.addXmlResponseBodyCheck(blockBuilder, xml, matchers, xml, true);
+		// Then
+		String test = blockBuilder.toString();
+		assertThat(test).contains(
```
Just FYI you can chain those contains into one.

javiersvg replied:
done

marcingrzejszczak commented:
Thanks @javiersvg for this PR! Great work!
